File: utils/request.R
# Imports ----------------------------------------------------------------
library(magrittr)

# Constants --------------------------------------------------------------
BASE_URL <- "https://data.cityofchicago.org/resource"

# Boilerplate query strings ----------------------------------------------
construct_url <- function(base_url = BASE_URL, view_id) {
    url <- glue::glue("{base_url}/{view_id}.json")
    return(url)
}

within_circle <- function(field = "location", latitude, longitude, radius) {
    query_string <- glue::glue("within_circle({field}, {latitude}, {longitude}, {radius})")
    return(query_string)
}

get_previous_sunday <- function() {
    current_date <- Sys.time() %>% lubridate::with_tz("America/Chicago") %>% lubridate::as_date()
    previous_sunday <- lubridate::floor_date(current_date, "week")
    if (current_date == previous_sunday) previous_sunday <- previous_sunday - 7
    return(previous_sunday)
}

date_between <- function(field, previous_sunday) {
    following_sunday <- previous_sunday + 7
    query_string <- glue::glue("{field} >= '{previous_sunday}T00:00:00.000' AND {field} < '{following_sunday}T00:00:00.000'")
    return(query_string)
}

value_in <- function(field, values) {
    values_sql <- paste0("('", paste0(values, collapse = "', '"),"')")
    query_string <- glue::glue("{field} IN {values_sql}")
    return(query_string)
}

construct_where_string <- function(...) {
    strings <- list(...)
    where_string <- paste0(strings, collapse = " AND ")
    return(where_string)
}

# Check whether response contains zero records ---------------------------
check_empty_response <- function(response) {
    return(!"data.frame" %in% class(response) || nrow(response) == 0)
}

# Get business licenses --------------------------------------------------
get_business_licenses <- function(application_type = "ISSUE", search_area) {
    where_location <- within_circle(
        latitude = search_area$latitude,
        longitude = search_area$longitude,
        radius = search_area$radius
    )

    where_date <- date_between(
        field = "date_issued",
        previous_sunday = get_previous_sunday()
    )

    where_application_type <- value_in(
        field = "application_type",
        values = application_type
    )

    query <- list(
        `$where` = construct_where_string(
            where_location,
            where_date,
            where_application_type
        )
    )

    url <- construct_url(view_id = "uupf-x98q")

    r <- httr::GET(url = url, query = query)
    httr::stop_for_status(r)
    response <- r %>%
        httr::content(as = "text", encoding = "UTF-8") %>%
        jsonlite::fromJSON()

    if (check_empty_response(response)) return(response)

    licenses <- response %>%
        dplyr::arrange(dplyr::desc(date_issued), dplyr::desc(expiration_date)) %>%
        dplyr::mutate(
            collapsed_name = toupper(ifelse(legal_name == doing_business_as_name, legal_name, glue::glue("{legal_name} (DBA: {doing_business_as_name})"))),
            license_start_date = strtrim(license_start_date, 10),
            expiration_date = strtrim(expiration_date, 10)
        ) %>%
        dplyr::select(
            `Business Name` = collapsed_name,
            `Address` = address,
            `Start Date` = license_start_date,
            `End Date` = expiration_date,
            `License Type` = business_activity
        )

    return(licenses)
}

# Get food inspection results --------------------------------------------
get_food_inspections <- function(facility_type = "Restaurant", search_area) {
    where_center <- rgeos::readWKT(glue::glue("POINT ({search_area$longitude} {search_area$latitude})"))
    where_radius <- rgeos::gBuffer(where_center, width = search_area$radius / 111139)

    where_date <- date_between(
        field = "inspection_date",
        previous_sunday = get_previous_sunday()
    )

    where_facility_type <- value_in(
        field = "facility_type",
        values = facility_type
    )

    query <- list(
        `$where` = construct_where_string(
            where_date,
            where_facility_type
        )
    )

    url <- construct_url(view_id = "4ijn-s7e5")

    r <- httr::GET(url = url, query = query)
    httr::stop_for_status(r)
    response <- r %>%
        httr::content(as = "text", encoding = "UTF-8") %>%
        jsonlite::fromJSON()

    if (check_empty_response(response)) return(response)

    inspections <- response %>%
        dplyr::rowwise() %>%
        dplyr::mutate(
            within_radius = rgeos::gContains(
                where_radius,
                rgeos::readWKT(glue::glue("POINT ({longitude} {latitude})"))
            )
        ) %>%
        dplyr::ungroup() %>%
        dplyr::filter(within_radius) %>%
        dplyr::arrange(dplyr::desc(inspection_date)) %>%
        dplyr::mutate(
            collapsed_name = toupper(ifelse(dba_name == aka_name, dba_name, glue::glue("{dba_name} (AKA: {aka_name})"))),
            inspection_date = strtrim(inspection_date, 10)
        ) %>%
        dplyr::select(
            `Business Name` = collapsed_name,
            `Address` = address,
            `Inspection Date` = inspection_date,
            `Inspection Type` = inspection_type,
            `Result` = results,
            `Risk Level` = risk,
            `Violations` = violations
        )

    return(inspections)
}

# Get filming permits ----------------------------------------------------
get_filming_permits <- function(search_area) {
    where_location <- within_circle(
        latitude = search_area$latitude,
        longitude = search_area$longitude,
        radius = search_area$radius
    )

    where_date <- date_between(
        field = "applicationissueddate",
        previous_sunday = get_previous_sunday()
    )

    query <- list(
        `$where` = construct_where_string(
            where_location,
            where_date
        )
    )

    url <- construct_url(view_id = "c2az-nhru")

    r <- httr::GET(url = url, query = query)
    httr::stop_for_status(r)
    response <- r %>%
        httr::content(as = "text", encoding = "UTF-8") %>%
        jsonlite::fromJSON()

    if (check_empty_response(response)) return(response)

    permits <- response %>%
        dplyr::arrange(applicationstartdate, applicationenddate) %>%
        dplyr::mutate(
            primarycontactlast = toupper(primarycontactlast),
            street_numbers = glue::glue("{streetnumberfrom}-{streetnumberto}"),
            applicationstartdate = strtrim(applicationstartdate, 10),
            applicationenddate = lubridate::ymd(strtrim(applicationenddate, 10)) + 1
        ) %>%
        tidyr::unite("address", street_numbers, direction, streetname, suffix, na.rm = TRUE, sep = " ") %>%
        dplyr::select(
            `Contact Name` = primarycontactlast,
            `Address` = address,
            `Start Date` = applicationstartdate,
            `End Date` = applicationenddate,
            `Details` = detail
        )

    return(permits)
}

File: chi_updates/request.py
from abc import ABC
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import List, Optional

import numpy as np
import pandas as pd
from sodapy import Socrata

SOCRATA_URL = "data.cityofchicago.org"


class SocrataDatasets(ABC):
    """
    Contains constants representing Socrata dataset IDs for various open
    datasets.

    Attributes:
        BUSINESS_LICENSES (str): Dataset ID for business licenses.
        FOOD_INSPECTIONS (str): Dataset ID for food inspections.
        FILMING_PERMITS (str): Dataset ID for filming permits.
        BUILDING_PERMITS (str): Dataset ID for building permits.

    """

    BUSINESS_LICENSES = "uupf-x98q"
    FOOD_INSPECTIONS = "4ijn-s7e5"
    FILMING_PERMITS = "c2az-nhru"
    BUILDING_PERMITS = "ydr8-5enu"


@dataclass
class SearchArea:
    """
    Represents a geographic search area.

    Attributes:
        lat (float): The latitude of the center point.
        lng (float): The longitude of the center point.
        radius (int): The radius in meters to search around the center point.

    """

    lat: float
    lng: float
    radius: int


def _init_client() -> Socrata:
    """
    Initializes a Socrata client object.

    Returns:
        Socrata: Socrata client instance.

    """
    return Socrata(SOCRATA_URL, None)


def _within_circle(search_area: SearchArea, field: str = "location") -> str:
    """
    Generates a within circle filter string.

    Args:
        search_area (SearchArea): The search area to filter on.
        field (str, optional): The location field name. Defaults to "location".

    Returns:
        str: The filter string.

    """
    return f"within_circle({field}, {search_area.lat}, {search_area.lng}, {search_area.radius})"


def _date_between(field: str, last_n_days: int = 7) -> str:
    """
    Generates a date range filter string for the last/next N days.

    Args:
        field (str): The date field name.
        last_n_days (int, optional): The number of days. Defaults to 7.

    Returns:
        str: The filter string.

    """
    current_datetime = datetime.now()
    n_days_ago = current_datetime - timedelta(days=last_n_days)

    current_datetime_str = current_datetime.strftime("%Y-%m-%dT%H:%M:%S.000")
    n_days_ago_str = n_days_ago.strftime("%Y-%m-%dT00:00:00.000")

    min_datetime = min(n_days_ago_str, current_datetime_str)
    max_datetime = max(n_days_ago_str, current_datetime_str)

    return f"{field} >= '{min_datetime}' AND {field} < '{max_datetime}'"


def _value_in(field: str, values: List[str], negate: bool = False) -> str:
    """
    Generates a IN filter string.

    Args:
        field (str): The field name.
        values (List[str]): The values to filter on.
        negate (bool, optional): Whether to negate the filter. Defaults to False.

    Returns:
        str: The filter string.

    """
    values_str = "', '".join(values)
    return f"{field} {'NOT' if negate else ''} IN ('{values_str}')"


def get_business_licenses(
    search_area: SearchArea,
    application_type: List[str],
) -> Optional[pd.DataFrame]:
    """
    Fetches business license data from Socrata.

    Args:
        search_area (SearchArea): The geographic area to search within.
        application_type (List[str]): Filter by application type.

    Returns:
        Optional[pd.DataFrame]: A DataFrame of business license records,
            or None if no records found.

    """
    # Construct where clause
    where_location = _within_circle(search_area=search_area)
    where_date = _date_between(field="date_issued")
    where_application_type = _value_in(
        field="application_type", values=application_type
    )
    where = " AND ".join([where_location, where_date, where_application_type])

    # Retrieve results
    client = _init_client()
    results = client.get_all(SocrataDatasets.BUSINESS_LICENSES, where=where)
    results_df = pd.DataFrame.from_records(results)

    if len(results_df) > 0:
        # Concatenate legal and DBA names
        results_df["Business Name"] = np.where(
            results_df["legal_name"] == results_df["doing_business_as_name"],
            results_df["legal_name"],
            results_df.apply(
                lambda x: f"{x['legal_name']} (DBA: {x['doing_business_as_name']})",
                axis=1,
            ).values,
        )

        # Trim date column(s)
        results_df["Start Date"] = results_df["license_start_date"].str[0:10]
        results_df["End Date"] = results_df["expiration_date"].str[0:10]

        # Rename remaining fields
        results_df = results_df.rename(
            columns={"address": "Address", "business_activity": "License Type"}
        )

        # Select relevant columns and sort
        results_df = results_df.loc[
            :,
            results_df.columns.isin(
                ["Business Name", "Address", "Start Date", "End Date", "License Type"]
            ),
        ]
        results_df = results_df.sort_values(
            by=["Start Date", "End Date"], ascending=False
        )

        return results_df

    else:
        return None


def get_food_inspections(
    search_area: SearchArea,
    facility_type: List[str],
    results: List[str],
) -> Optional[pd.DataFrame]:
    """
    Fetches food inspection data from Socrata.

    Args:
        search_area (SearchArea): The geographic area to search within.
        facility_type (List[str]): Filter by facility type.
        results (List[str]): Filter by inspection result.

    Returns:
        Optional[pd.DataFrame]: A DataFrame of food inspection records,
            or None if no records found.

    """
    # Construct where clause
    where_location = _within_circle(search_area=search_area)
    where_date = _date_between(field="inspection_date")
    where_facility_type = _value_in(field="facility_type", values=facility_type)
    where_results = _value_in(field="results", values=results)
    where = " AND ".join(
        [where_location, where_date, where_facility_type, where_results]
    )

    # Retrieve results
    client = _init_client()
    results = client.get_all(SocrataDatasets.FOOD_INSPECTIONS, where=where)
    results_df = pd.DataFrame.from_records(results)

    if len(results_df) > 0:
        # Concatenate legal and DBA names
        results_df["Business Name"] = np.where(
            results_df["dba_name"] == results_df["aka_name"],
            results_df["dba_name"],
            results_df.apply(
                lambda x: f"{x['dba_name']} (AKA: {x['aka_name']})", axis=1
            ).values,
        )

        # Trim date column(s)
        results_df["Inspection Date"] = results_df["inspection_date"].str[0:10]

        # Create violations column if not exists
        results_df["Violations"] = (
            results_df["violations"] if "violations" in results_df.columns else None
        )

        # Rename remaining fields
        results_df = results_df.rename(
            columns={
                "address": "Address",
                "inspection_type": "Inspection Type",
                "results": "Results",
                "risk": "Risk Level",
            }
        )

        # Select relevant columns and sort
        results_df = results_df.loc[
            :,
            results_df.columns.isin(
                [
                    "Business Name",
                    "Address",
                    "Inspection Date",
                    "Inspection Type",
                    "Results",
                    "Risk Level",
                    "Violations",
                ]
            ),
        ]
        results_df = results_df.sort_values(
            by=["Inspection Date", "Business Name"], ascending=False
        )

        return results_df

    else:
        return None


def get_filming_permits(
    search_area: SearchArea,
    application_status: List[str],
) -> Optional[pd.DataFrame]:
    """
    Fetches filming permit data from Socrata.

    Args:
        search_area (SearchArea): The geographic area to search within.
        application_status (List[str]): Filter by application status.

    Returns:
        Optional[pd.DataFrame]: A DataFrame of filming permit records,
            or None if no records found.

    """
    # Construct where clause
    where_location = _within_circle(search_area=search_area)
    where_date = _date_between(field="applicationstartdate", last_n_days=-7)
    where_application_status = _value_in(
        field="currentmilestone", values=application_status, negate=True
    )
    where = " AND ".join([where_location, where_date, where_application_status])

    # Retrieve results
    client = _init_client()
    results = client.get_all(SocrataDatasets.FILMING_PERMITS, where=where)
    results_df = pd.DataFrame.from_records(results)

    if len(results_df) > 0:
        # Concatenate address components
        results_df["street_numbers"] = results_df.apply(
            lambda x: f"{x['streetnumberfrom']}-{x['streetnumberto']}", axis=1
        ).values
        results_df["Address"] = results_df.apply(
            lambda x: x[
                ["street_numbers", "direction", "streetname", "suffix"]
            ].str.cat(sep=" "),
            axis=1,
        )

        # Trim date column(s)
        results_df["Start Date"] = results_df["applicationstartdate"].str[0:10]
        results_df["End Date"] = results_df["applicationenddate"].str[0:10]

        # Rename remaining fields
        results_df = results_df.rename(
            columns={
                "primarycontactlast": "Contact Name",
                "applicationname": "Application Name",
                "detail": "Details",
                "comments": "Comments",
            }
        )

        # Select relevant columns and sort
        results_df = results_df.loc[
            :,
            results_df.columns.isin(
                [
                    "Contact Name",
                    "Application Name",
                    "Address",
                    "Start Date",
                    "End Date",
                    "Details",
                    "Comments",
                ]
            ),
        ]
        results_df = results_df.sort_values(by=["Start Date", "End Date"])

        return results_df

    else:
        return None
